### PR TITLE
add sum query problem

### PR DESCRIPTION
**Pull request by RaduNichita**
add sum_query problem

- examples
-   C++ implementation

### CODE REVIEW

File: lab/dp/problems/sum_query/enunt.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Gigel s-a lasat de stropitori si de mancat pufuleti, asa ca s-a apucat de 
+carte si a inceput fix cu ce era mai greu. Asa ca el va cere voua ajutorul,
+ca sa poate spune prietenilor lui ca e destept. Gigel va da o matrice N * M,
+cu numere intregi. El se intreaba cum poate calcula extrem de rapid(cu viteza
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/enunt.md:4]:
```suggestion
cu numere intregi. El se intreaba cum poate calcula extrem de rapid (cu viteza
```

File: lab/dp/problems/sum_query/enunt.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+tunetului si a fulgerului) suma dintr-un patrat oarecare din matrice(determinat
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/enunt.md:5]:
```suggestion
tunetului si a fulgerului) suma dintr-un patrat oarecare din matrice (determinat
```

File: lab/dp/problems/sum_query/enunt.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+de coordonatele initiale xi si yi, si coordonatele finale xf si yf). El 
+primeste un numar mare de intrebari de genul "Care este suma elementelor 
+patratului determinat de coordonatele (X, Y) si (Z, T). de la prietenii sai, 
+si doreste sa le arate ca el acum este foarte destept si ca aspira la premiul 1
+in clasa.
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/enunt.md:10]:
```suggestion
de coordonatele initiale xi si yi, si coordonatele finale xf si yf). El 
primeste un numar mare de intrebari de forma:

Care este suma elementelor  patratului determinat de coordonatele (X, Y) si (Z, T) de la prietenii sai. El doreste sa le arate ca stie o metoda eficienta pentru a le raspunde la intrebari.
```

File: lab/dp/problems/sum_query/main.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    std::vector<int> solution;
+
+    void read_input() {
+        std::ifstream in("sum_query.in");
+        in >> N;
+        in >> M;
+        in >> K;
+        matrix =
+        std::vector<std::vector<int>>(N + 1, std::vector<int>(M + 1, 0));
+        dp =
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:33]: Did you run clang-format? I bet on not!

File: lab/dp/problems/sum_query/main.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        for (int j = 1; j <= M; j++) {
+            for (int i = 1; i <= N; i++) {
+                dp[i][j] += dp[i - 1][j];
+            }
+        }
+    }
+
+    void show_matrix(std::vector<std::vector<int>> matrix) {
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:84]:
```suggestion
    void show_matrix(const std::vector<std::vector<int>>& matrix) {
```

File: lab/dp/problems/sum_query/main.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    // |   +------+    |   |         |    |   |   |           |   +---------+    |   +---+          |
+    // |   |      |    | = |         |    | - |   |           | - |      (r1,c2) | + |   (r1,c1)    |
+    // |   |      |    |   |         |    |   |   |           |   |              |   |              |
+    // |   +------+    |   +---------+    |   +---+           |   |              |   |              |
+    // |        (r2,c2)|   |       (r2,c2)|   |   (r2,c1)     |   |              |   |              |
+    // +---------------+   +--------------+   +---------------+   +--------------+   +--------------+
+    void get_result() {
+        compute_sum_matrix();
+        for (int i = 1; i <= K; i++) {
+            auto start = queries[i].first;
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:110]: If you use C++17, you have structured bindings (aka for (const auto& [start, end] : queries]) )

File: lab/dp/problems/sum_query/main.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            sol = sol - dp[start.first - 1][end.second];
+            sol = sol - dp[end.first][start.second - 1];
+            sol = sol + dp[start.first - 1][start.second - 1];
+            solution.push_back(sol);
+        }
+    }
+
+    void print_output() {
+        std::ofstream out("sum_query.out");
+        for (auto const &sol : solution) {
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:122]:
```suggestion
        for (const auto &sol : solution) {
```

File: lab/dp/problems/sum_query/main.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+ public:
+    void solve() {
+        read_input();
+        get_result();
+        print_output();
+    }
+};
+
+int main() {
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:136]: Please use the main from template.cpp. You can comment redirection if this is tested on leetcode.

File: lab/dp/problems/sum_query/main.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    std::vector<std::vector<int>> dp;
+
+    // vectorul se solutii
+    std::vector<int> solution;
+
+    void read_input() {
+        std::ifstream in("sum_query.in");
+        in >> N;
+        in >> M;
+        in >> K;
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:30]:
```suggestion
        in >> N >> M >> K;
```

File: lab/dp/problems/sum_query/main.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>> queries;
+
+    // pe pozitia [i][j] am suma elementelor submatricei (1,1) pana la (i, j).
+    std::vector<std::vector<int>> dp;
+
+    // vectorul se solutii
+    std::vector<int> solution;
+
+    void read_input() {
+        std::ifstream in("sum_query.in");
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:27]: Please use cin and cout as in template.cpp

File: lab/dp/problems/sum_query/main.cpp (hunk 4 above)
Comment (RaduNichita) [reply lab/dp/problems/sum_query/main.cpp]: I had some trouble, but a purge and a clean install did the trick as always :)

**Comment by RaduNichita** [comment]
> Please fix reported issues

Now is ok, @neatudarius ?

File: lab/dp/problems/sum_query/enunt.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Gigel si suma elementelor din matrice
+
+*Gigel* s-a lasat de stropitori si de mancat pufuleti, asa ca s-a apucat de 
+carte si a inceput fix cu ce era mai greu. Asa ca el va cere voua ajutorul,
+ca sa poate spune prietenilor lui ca e destept. Gigel va da o matrice N * M,
+cu numere intregi. El se intreaba cum poate calcula extrem de rapid (cu viteza
+tunetului si a fulgerului) suma dintr-un patrat oarecare din matrice (determinat
+de coordonatele initiale xi si yi, si coordonatele finale xf si yf). El
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/enunt.md:8]: I thinl `initiale` and `finale` are not proper terms. We should use "stanga sus" si "dreapta jos" (top-left and bottom-right)

File: lab/dp/problems/sum_query/main.cpp [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    // matricea de elemente
+    std::vector<std::vector<int>> matrix;
+
+    // query-urile de forma <(start_x, start,y), (finish_x), (finish_y)>.
+    std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>> queries;
+
+    // pe pozitia [i][j] am suma elementelor submatricei (1,1) pana la (i, j).
+    std::vector<std::vector<int>> dp;
+
+    // vectorul se solutii
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:23]:
```suggestion
    // vectorul de solutii
```

File: lab/dp/problems/sum_query/main.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+// SumQuery - Programare dinamica
+// Complexitate O(M * N)
+
+// https://leetcode.com/problems/range-sum-query-2d-immutable/
+
+#include <bits/stdc++.h>
+
+class Task {
+    // N si M sunt dimensiunile matricei, K e numarul de query-uri
+    int N, M, K;
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:12]: please use lowercase for variables (and uppercase for macros)

File: lab/dp/problems/sum_query/sum_query.in [hunk 14]
```diff
@@ -0,0 +1,7 @@
+3 3 3
+1 2 3
+4 5 6
+7 8 9
+1 1 2 2 
+2 2 2 3
+3 2 3 2
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/sum_query.in:7]:
```suggestion
3 2 3 2

```

File: lab/dp/problems/sum_query/main.cpp [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    void show_matrix(const std::vector<std::vector<int>> matrix) {
+        for (auto const& line : matrix) {
+            for (auto const& elem : line) {
+                std::cout << elem << " ";
+            }
+            std::cout << "\n";
+        }
+    }
+
+    // pentru a calcula suma elementelor dintre colturile stanga-sus (r1,c1) si
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:86]: Please use the same corner notation in the entire directory


x/y z/t vs initial / final vs r1c1 r2 c2

I remind you that we should use a "standard" notation in this repo.

What do you think about (x,y) - (p, q) ? 

And i,j, k for loop counters.

File: lab/dp/problems/sum_query/main.cpp [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        for (int j = 1; j <= M; j++) {
+            for (int i = 1; i <= N; i++) {
+                dp[i][j] += dp[i - 1][j];
+            }
+        }
+    }
+
+    void show_matrix(const std::vector<std::vector<int>> matrix) {
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:77]:
```suggestion
    void show_matrix(const std::vector<std::vector<int>>& matrix) {
```

File: lab/dp/problems/sum_query/main.cpp [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        matrix = std::vector<std::vector<int>>(N + 1, std::vector<int>(M + 1, 0));
+        dp     = std::vector<std::vector<int>>(N + 1, std::vector<int>(M + 1, 0));
+
+        for (int i = 1; i <= M; i++) {
+            for (int j = 1; j <= N; j++) {
+                std::cin >> matrix[i][j];
+            }
+        }
+
+        queries = std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>(K);
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:37]: I found this very confusion. Please use `push_back` in the loop and avoid this ugly code.

`queries.push_back({ {x, y}, {p, q} };` - this whould make the code mode readable as you don't need the line 37.

File: lab/dp/problems/sum_query/main.cpp [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+    // pe pozitia [i][j] am suma elementelor submatricei (1,1) pana la (i, j).
+    std::vector<std::vector<int>> dp;
+
+    // vectorul se solutii
+    std::vector<int> solution;
+
+    void read_input() {
+        std::cin >> N >> M >> K;
+        matrix = std::vector<std::vector<int>>(N + 1, std::vector<int>(M + 1, 0));
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:28]: Not a todo for this PR: what do you think to make a mini-library with useful wrappers / typedefs/ etc?

E.g.

using matrix_t =  std::vector<std::vector<int>>;
#define build_matrix(n, m, initial_value) std::vector<std::vector<int>>(n + 1, std::vector<int>(m + 1, inital_value))

We can think as a todo fot the future if this abstraction will help the students.

File: lab/dp/problems/sum_query/main.cpp [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+    // N si M sunt dimensiunile matricei, K e numarul de query-uri
+    int N, M, K;
+
+    // matricea de elemente
+    std::vector<std::vector<int>> matrix;
+
+    // query-urile de forma <(start_x, start,y), (finish_x), (finish_y)>.
+    std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>> queries;
+
+    // pe pozitia [i][j] am suma elementelor submatricei (1,1) pana la (i, j).
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:20]:
```suggestion
    // dp[i][j] = suma elementelor submatricei (1,1) - (i, j).
```

File: lab/dp/problems/sum_query/main.cpp [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+// https://leetcode.com/problems/range-sum-query-2d-immutable/
+
+#include <bits/stdc++.h>
+
+class Task {
+    // N si M sunt dimensiunile matricei, K e numarul de query-uri
+    int N, M, K;
+
+    // matricea de elemente
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:14]:
```suggestion
    // matricea citita
```

File: lab/dp/problems/sum_query/main.cpp [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    std::vector<std::vector<int>> matrix;
+
+    // query-urile de forma <(start_x, start,y), (finish_x), (finish_y)>.
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:17]:
```suggestion
    // query-urile de forma <(start_x, start,y), (finish_x, (finish_y)>.
```

These 4 values also need to be renamed.

File: lab/dp/problems/sum_query/main.cpp [hunk 22]
```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 Nichita Radu <radunichita99@gmail.com>
+
+// SumQuery - Programare dinamica
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:3]:
```suggestion
// SumQuery - DP
```

File: lab/dp/problems/sum_query/main.cpp [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+// Complexitate O(M * N)
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:4]:
```suggestion
// Complexitate O(n * m)
```

File: lab/dp/problems/sum_query/example [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+7 8 9, 
+matricea intermediara formata din prin adunarea elementului curent la 
+precedentul de pe linie va fi:
+1 3 6
+4 9 15
+7 15 24, iar matricea dp se va obtine la adunarea elementului curent la 
+precedentul de pe coloana, adica:
+1 3 6
+5 12 21
+12 27 45
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/example:13]:
```suggestion
12 27 45

```

File: lab/dp/problems/sum_query/Makefile [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+
+build:
+	$(CC) $(CFLAGS) $(LDFLAGS) $(SRC) -o $(EXE) 
+
+run: build
+	time ./$(EXE) && echo "pwp" &> /dev/null
+
+.PHONY: clean
+clean:
+	rm -f *.o *.exe *.out
```
Comment (neatudarius) [inline lab/dp/problems/sum_query/Makefile:18]: You should always have a newline at the end of a file.
```suggestion
	rm -f *.o *.exe *.out

```

File: lab/dp/problems/sum_query/main.cpp (hunk 21 above)
Comment (neatudarius) [inline lab/dp/problems/sum_query/main.cpp:17]:
```suggestion
    // query-uri de forma <(start_x, start,y), (finish_x), (finish_y)>.
```

**Review by neatudarius — APPROVED** [review]

**Pull request merged**